<a href="https://colab.research.google.com/github/swaroopa1352/NNSwaroopa/blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [16]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipython-input-16-1433581267.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633101  129.181519  125.971433  127.555151  1.120920e+08
std      30.421461   30.698898   30.139548   30.449561  5.602586e+07
min      54.378582   55.379535   51.528416   55.277744  3.519590e+07
25%     112.709538  114.204933  111.251818  112.889687  7.636470e+07
50%     133.010765  134.143878  131.050657  132.832284  9.493580e+07
75%     148.221512  149.388600  146.276903  147.588785  1.296327e+08
max     178.645645  179.558457  175.809061  179.254190  4.265100e+08


In [17]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipython-input-17-3699561310.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633101  129.181519  125.971433  127.555151  1.120920e+08
std      30.421461   30.698898   30.139548   30.449561  5.602586e+07
min      54.378582   55.379535   51.528416   55.277744  3.519590e+07
25%     112.709538  114.204933  111.251818  112.889687  7.636470e+07
50%     133.010765  134.143878  131.050657  132.832284  9.493580e+07
75%     148.221512  149.388600  146.276903  147.588785  1.296327e+08
max     178.645645  179.558457  175.809061  179.254190  4.265100e+08


In [18]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
import pandas as pd # Ensure pandas is imported
import yfinance as yf # Import yfinance

# Load historical data for AAPL again to ensure data is available
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Check if the data DataFrame is not empty after loading
if not data.empty:

    # Feature Engineering

    # Moving Averages
    data['MA_5'] = data['Close'].rolling(window=5).mean()
    data['MA_10'] = data['Close'].rolling(window=10).mean()
    data['MA_50'] = data['Close'].rolling(window=50).mean()

    # Volatility Measures
    data['Vol_Change'] = data['Volume'].pct_change() # Percentage change in Volume
    data['Price_Change'] = data['Close'].pct_change() # Percentage change in Price
    data['Daily_Return'] = data['Close'].pct_change()
    data['Volatility'] = data['Daily_Return'].rolling(window=14).std() # 14-day rolling standard deviation

    # Lagged Prices
    data['Close_Lag1'] = data['Close'].shift(1)
    data['Close_Lag5'] = data['Close'].shift(5)

    # Manually calculate a simple technical indicator (e.g., Relative Strength Index - RSI)
    # This is a simplified RSI calculation and may differ from pandas_ta
    delta = data['Close'].diff()
    gain = delta.mask(delta < 0, 0)
    loss = delta.mask(delta > 0, 0).abs()
    avg_gain = gain.ewm(com=13, adjust=False).mean()
    avg_loss = loss.ewm(com=13, adjust=False).mean()
    rs = avg_gain / avg_loss
    data['RSI_14_manual'] = 100 - (100 / (1 + rs))


    # Create the target variable: tomorrow's closing price
    data['Target'] = data['Close'].shift(-1)

    # Inspect the data after feature engineering but before dropping NaN values
    print("Data after feature engineering but before dropping NaN values:")
    display(data.head())
    display(data.tail())
    print("\nShape of data after feature engineering but before dropping NaN values:", data.shape)
    print("Number of NaN values per column after feature engineering but before dropping:")
    display(data.isnull().sum())

    # Drop rows with NaN values created by feature engineering and target variable
    data.dropna(inplace=True)

    # Inspect the data after dropping NaN values
    print("\nData after dropping NaN values:")
    display(data.head())
    display(data.tail())
    print("\nShape of data after dropping NaN values:", data.shape)

    # Check if data is still not empty after dropping NaN values
    if not data.empty:
        # Define features (X) and target (y) - include new features
        features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50',
                    'Vol_Change', 'Price_Change', 'Daily_Return', 'Volatility',
                    'Close_Lag1', 'Close_Lag5', 'RSI_14_manual']
        X = data[features]
        y = data['Target']

        # Split data into training and testing sets (70/30 ratio)
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

        # Initialize and train the Linear Regression model
        model = LinearRegression()
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Evaluate the model
        mse = mean_squared_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)

        print(f"\nMean Squared Error (MSE): {mse:.2f}")
        print(f"R-squared (R²) Score: {r2:.2f}")

    else:
        print("\nData is empty after dropping NaN values. Cannot proceed with model training.")

else:
    print("The 'data' DataFrame was empty after loading. Please check the yfinance download step.")

/tmp/ipython-input-18-1014061442.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Data after feature engineering but before dropping NaN values:


Price           Close       High        Low       Open     Volume       MA_5  \
Ticker           AAPL       AAPL       AAPL       AAPL       AAPL              
Date                                                                           
2020-01-02  72.620834  72.681281  71.373211  71.627084  135480400        NaN   
2020-01-03  71.914818  72.676447  71.689957  71.847118  146322800        NaN   
2020-01-06  72.487854  72.526541  70.783256  71.034717  118387200        NaN   
2020-01-07  72.146950  72.753831  71.926922  72.497537  108872000        NaN   
2020-01-08  73.307518  73.609752  71.849540  71.849540  132079200  72.495595   

Price      MA_10 MA_50 Vol_Change Price_Change Daily_Return Volatility  \
Ticker                                                                   
Date                                                                     
2020-01-02   NaN   NaN        NaN          NaN          NaN        NaN   
2020-01-03   NaN   NaN   0.080029    -0.009722    -0.009722        NaN   
2020-01-06   NaN   NaN  -0.190918     0.007968     0.007968        NaN   
2020-01-07   NaN   NaN  -0.080374    -0.004703    -0.004703        NaN   
2020-01-08   NaN   NaN   0.213160     0.016086     0.016086        NaN   

Price      Close_Lag1 Close_Lag5 RSI_14_manual     Target  
Ticker                                                     
Date                                                       
2020-01-02        NaN        NaN           NaN  71.914818  
2020-01-03  72.620834        NaN      0.000000  72.487854  
2020-01-06  71.914818        NaN      5.876540  72.146950  
2020-01-07  72.487854        NaN      5.663320  73.307518  
2020-01-08  72.146950        NaN     16.739008  74.864609

Price            Close        High         Low        Open    Volume  \
Ticker            AAPL        AAPL        AAPL        AAPL      AAPL   
Date                                                                   
2022-12-23  130.173782  130.726619  127.982170  129.245801  63814900   
2022-12-27  128.367203  129.729560  127.073957  129.699945  69007800   
2022-12-28  124.428215  129.354401  124.260391  128.011792  85438400   
2022-12-29  127.952568  128.811438  126.096612  126.353282  75703700   
2022-12-30  128.268478  128.288227  125.800455  126.767927  77034200   

Price             MA_5       MA_10       MA_50 Vol_Change Price_Change  \
Ticker                                                                   
Date                                                                     
2022-12-23  131.143225  135.089108  142.216760  -0.180306    -0.002798   
2022-12-27  130.681216  133.661601  142.056421   0.081374    -0.013878   
2022-12-28  129.445229  131.743449  141.737864   0.238098    -0.030685   
2022-12-29  128.292166  130.400842  141.463382  -0.113938     0.028324   
2022-12-30  127.838049  129.752246  141.193049   0.017575     0.002469   

Price      Daily_Return Volatility  Close_Lag1  Close_Lag5 RSI_14_manual  \
Ticker                                                                     
Date                                                                       
2022-12-23    -0.002798   0.018781  130.539062  132.789871     35.694643   
2022-12-27    -0.013878   0.018176  130.173782  130.677246     33.675348   
2022-12-28    -0.030685   0.019223  128.367203  130.608154     29.726642   
2022-12-29     0.028324   0.020925  124.428215  133.717880     36.860483   
2022-12-30     0.002469   0.021053  127.952568  130.539062     37.473212   

Price           Target  
Ticker                  
Date                    
2022-12-23  128.367203  
2022-12-27  124.428215  
2022-12-28  127.952568  
2022-12-29  128.268478  
2022-12-30         NaN


Shape of data after feature engineering but before dropping NaN values: (756, 16)
Number of NaN values per column after feature engineering but before dropping:


Price          Ticker
Close          AAPL       0
High           AAPL       0
Low            AAPL       0
Open           AAPL       0
Volume         AAPL       0
MA_5                      4
MA_10                     9
MA_50                    49
Vol_Change                1
Price_Change              1
Daily_Return              1
Volatility               14
Close_Lag1                1
Close_Lag5                5
RSI_14_manual             1
Target                    1
dtype: int64


Data after dropping NaN values:


Price           Close       High        Low       Open     Volume       MA_5  \
Ticker           AAPL       AAPL       AAPL       AAPL       AAPL              
Date                                                                           
2020-03-13  67.369156  67.841763  61.305276  64.199074  370732000  65.589732   
2020-03-16  58.702312  62.790941  58.166693  58.639296  322423600  64.428338   
2020-03-17  61.283459  62.434671  57.778914  59.986825  324056000  62.853963   
2020-03-18  59.783234  60.590297  57.468684  58.110943  300233600  61.459897   
2020-03-19  59.325180  61.278611  58.799257  59.957743  271857200  61.292668   

Price           MA_10      MA_50 Vol_Change Price_Change Daily_Return  \
Ticker                                                                  
Date                                                                    
2020-03-13  68.490313  74.025908  -0.114086     0.119808     0.119808   
2020-03-16  67.118549  73.747538  -0.130305    -0.128647    -0.128647   
2020-03-17  66.234898  73.534911   0.005063     0.043970     0.043970   
2020-03-18  64.875979  73.280818  -0.073513    -0.024480    -0.024480   
2020-03-19  63.709251  73.024383  -0.094514    -0.007662    -0.007662   

Price      Volatility Close_Lag1 Close_Lag5 RSI_14_manual     Target  
Ticker                                                                
Date                                                                  
2020-03-13   0.065253  60.161324  70.049667     44.954384  58.702312  
2020-03-16   0.073137  67.369156  64.509285     36.656287  61.283459  
2020-03-17   0.074273  58.702312  69.155334     40.196808  59.783234  
2020-03-18   0.072621  61.283459  66.753563     38.838027  59.325180  
2020-03-19   0.072614  59.783234  60.161324     38.411095  55.558891

Price            Close        High         Low        Open    Volume  \
Ticker            AAPL        AAPL        AAPL        AAPL      AAPL   
Date                                                                   
2022-12-22  130.539062  132.839269  128.633750  132.631963  77852100   
2022-12-23  130.173782  130.726619  127.982170  129.245801  63814900   
2022-12-27  128.367203  129.729560  127.073957  129.699945  69007800   
2022-12-28  124.428215  129.354401  124.260391  128.011792  85438400   
2022-12-29  127.952568  128.811438  126.096612  126.353282  75703700   

Price             MA_5       MA_10       MA_50 Vol_Change Price_Change  \
Ticker                                                                   
Date                                                                     
2022-12-22  131.666443  136.105939  142.431838  -0.093984    -0.023773   
2022-12-23  131.143225  135.089108  142.216760  -0.180306    -0.002798   
2022-12-27  130.681216  133.661601  142.056421   0.081374    -0.013878   
2022-12-28  129.445229  131.743449  141.737864   0.238098    -0.030685   
2022-12-29  128.292166  130.400842  141.463382  -0.113938     0.028324   

Price      Daily_Return Volatility  Close_Lag1  Close_Lag5 RSI_14_manual  \
Ticker                                                                     
Date                                                                       
2022-12-22    -0.023773   0.018734  133.717880  134.754440     36.101079   
2022-12-23    -0.002798   0.018781  130.539062  132.789871     35.694643   
2022-12-27    -0.013878   0.018176  130.173782  130.677246     33.675348   
2022-12-28    -0.030685   0.019223  128.367203  130.608154     29.726642   
2022-12-29     0.028324   0.020925  124.428215  133.717880     36.860483   

Price           Target  
Ticker                  
Date                    
2022-12-22  130.173782  
2022-12-23  128.367203  
2022-12-27  124.428215  
2022-12-28  127.952568  
2022-12-29  128.268478


Shape of data after dropping NaN values: (706, 16)

Mean Squared Error (MSE): 6.91
R-squared (R²) Score: 0.99


In [28]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np
import pandas as pd

# Check if the data DataFrame from previous steps is not empty
# We will use the 'data' DataFrame which should contain the features engineered in the previous step
if 'data' in locals() and not data.empty:

    # Flatten the MultiIndex
    data.columns = ['_'.join(col).strip() if isinstance(col, tuple) else col for col in data.columns]

    # Print column names after flattening for debugging
    print("Column names after flattening:")
    display(data.columns)

    # Ensure 'Target_' column exists, recreating it at the top level if necessary using the flattened name
    if 'Target_' not in data.columns or data['Target_'].isnull().all(): # Check if 'Target_' is missing or all NaN
        # Check if 'Close_AAPL' exists after flattening
        if 'Close_AAPL' in data.columns:
            data['Target_'] = data['Close_AAPL'].shift(-1)
        else:
            print("Error: 'Close_AAPL' column not found in the DataFrame after flattening.")
            # Exit the execution of this cell if 'Close_AAPL' is not found
            raise SystemExit("Required column not found.")

        # Re-drop NaN values that might be created by the new Target_ column if it was missing
        data.dropna(inplace=True)


    # Create binary target based on whether tomorrow's close is higher than today's close
    # Access 'Close_AAPL' and 'Target_' at top level after flattening
    close_price_flattened = data['Close_AAPL']
    data['Price_Direction'] = (data['Target_'] > close_price_flattened).astype(int)

    # Drop the original 'Target_' column as we now have the binary 'Price_Direction'
    if 'Target_' in data.columns:
         data.drop('Target_', axis=1, inplace=True)


    # Inspect the data after creating the binary target
    print("Data after creating binary target variable:")
    display(data.head())
    display(data.tail())
    print("\nShape of data after creating binary target variable:", data.shape)
    print("Value counts for the binary target variable:")
    display(data['Price_Direction'].value_counts())


    # Define features (X) and target (y)
    # Exclude the 'Price_Direction' column itself from features
    # Features will be all columns except 'Price_Direction' and the flattened 'Target_'
    features = [col for col in data.columns if col not in ['Price_Direction', 'Target_']]
    X = data[features]
    y = data['Price_Direction']


    # Split data into training and testing sets (70/30 ratio)
    # Ensure we have enough samples after dropping NaNs
    if len(X) > 1: # Check if there's more than one sample to split
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y) # Use stratify for classification

        # Initialize and train the Random Forest Classifier model
        model = RandomForestClassifier(n_estimators=100, random_state=42) # Using 100 estimators as a starting point
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Evaluate the model
        accuracy = accuracy_score(y_test, y_pred)
        conf_matrix = confusion_matrix(y_test, y_pred)
        class_report = classification_report(y_test, y_pred)

        print(f"\nAccuracy: {accuracy:.2f}")
        print("\nConfusion Matrix:")
        display(conf_matrix)
        print("\nClassification Report:")
        print(class_report)

        # Report Feature Importance
        print("\nFeature Importance:")
        # Ensure feature_importances index matches the actual column names in X_train
        feature_importances = pd.Series(model.feature_importances_, index=X_train.columns).sort_values(ascending=False)
        display(feature_importances)

    else:
        print("\nNot enough data after creating binary target to perform train-test split.")


else:
    print("The 'data' DataFrame is empty. Please ensure the previous steps to load and process data were successful.")

Column names after flattening:


Index(['Close_AAPL', 'High_AAPL', 'Low_AAPL', 'Open_AAPL', 'Volume_AAPL',
       'MA_5_', 'MA_10_', 'MA_50_', 'Vol_Change_', 'Price_Change_',
       'Daily_Return_', 'Volatility_', 'Close_Lag1_', 'Close_Lag5_',
       'RSI_14_manual_', 'Price_Direction'],
      dtype='object')

Data after creating binary target variable:


Close_AAPL  High_AAPL   Low_AAPL  Open_AAPL  Volume_AAPL  \
Date                                                                   
2020-03-13   67.369156  67.841763  61.305276  64.199074    370732000   
2020-03-16   58.702312  62.790941  58.166693  58.639296    322423600   
2020-03-17   61.283459  62.434671  57.778914  59.986825    324056000   
2020-03-18   59.783234  60.590297  57.468684  58.110943    300233600   
2020-03-19   59.325180  61.278611  58.799257  59.957743    271857200   

                MA_5_     MA_10_     MA_50_  Vol_Change_  Price_Change_  \
Date                                                                      
2020-03-13  65.589732  68.490313  74.025908    -0.114086       0.119808   
2020-03-16  64.428338  67.118549  73.747538    -0.130305      -0.128647   
2020-03-17  62.853963  66.234898  73.534911     0.005063       0.043970   
2020-03-18  61.459897  64.875979  73.280818    -0.073513      -0.024480   
2020-03-19  61.292668  63.709251  73.024383    -0.094514      -0.007662   

            Daily_Return_  Volatility_  Close_Lag1_  Close_Lag5_  \
Date                                                               
2020-03-13       0.119808     0.065253    60.161324    70.049667   
2020-03-16      -0.128647     0.073137    67.369156    64.509285   
2020-03-17       0.043970     0.074273    58.702312    69.155334   
2020-03-18      -0.024480     0.072621    61.283459    66.753563   
2020-03-19      -0.007662     0.072614    59.783234    60.161324   

            RSI_14_manual_  Price_Direction  
Date                                         
2020-03-13       44.954384                0  
2020-03-16       36.656287                1  
2020-03-17       40.196808                0  
2020-03-18       38.838027                0  
2020-03-19       38.411095                0

Close_AAPL   High_AAPL    Low_AAPL   Open_AAPL  Volume_AAPL  \
Date                                                                      
2022-12-21  133.717880  135.060489  131.052411  131.279465     85928000   
2022-12-22  130.539062  132.839269  128.633750  132.631963     77852100   
2022-12-23  130.173782  130.726619  127.982170  129.245801     63814900   
2022-12-27  128.367203  129.729560  127.073957  129.699945     69007800   
2022-12-28  124.428215  129.354401  124.260391  128.011792     85438400   

                 MA_5_      MA_10_      MA_50_  Vol_Change_  Price_Change_  \
Date                                                                         
2022-12-21  132.509518  137.134613  142.547952     0.109711       0.023810   
2022-12-22  131.666443  136.105939  142.431838    -0.093984      -0.023773   
2022-12-23  131.143225  135.089108  142.216760    -0.180306      -0.002798   
2022-12-27  130.681216  133.661601  142.056421     0.081374      -0.013878   
2022-12-28  129.445229  131.743449  141.737864     0.238098      -0.030685   

            Daily_Return_  Volatility_  Close_Lag1_  Close_Lag5_  \
Date                                                               
2022-12-21       0.023810     0.018178   130.608154   141.378632   
2022-12-22      -0.023773     0.018734   133.717880   134.754440   
2022-12-23      -0.002798     0.018781   130.539062   132.789871   
2022-12-27      -0.013878     0.018176   130.173782   130.677246   
2022-12-28      -0.030685     0.019223   128.367203   130.608154   

            RSI_14_manual_  Price_Direction  
Date                                         
2022-12-21       39.759419                0  
2022-12-22       36.101079                0  
2022-12-23       35.694643                0  
2022-12-27       33.675348                0  
2022-12-28       29.726642                1


Shape of data after creating binary target variable: (705, 16)
Value counts for the binary target variable:


Price_Direction
1    364
0    341
Name: count, dtype: int64


Accuracy: 0.45

Confusion Matrix:


array([[41, 62],
       [54, 55]])


Classification Report:
              precision    recall  f1-score   support

           0       0.43      0.40      0.41       103
           1       0.47      0.50      0.49       109

    accuracy                           0.45       212
   macro avg       0.45      0.45      0.45       212
weighted avg       0.45      0.45      0.45       212


Feature Importance:


Volume_AAPL       0.080512
Volatility_       0.078213
RSI_14_manual_    0.075366
Price_Change_     0.073465
Daily_Return_     0.072254
Vol_Change_       0.070624
High_AAPL         0.065029
MA_50_            0.062870
Open_AAPL         0.062531
Close_Lag5_       0.061998
Close_AAPL        0.061401
Low_AAPL          0.060912
MA_5_             0.060216
Close_Lag1_       0.059148
MA_10_            0.055462
dtype: float64

In [27]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np
import pandas as pd

# Check if the data DataFrame from previous steps is not empty
# We will use the 'data' DataFrame which should contain the features engineered in the previous step
if 'data' in locals() and not data.empty:

    # Flatten the MultiIndex
    data.columns = ['_'.join(col).strip() if isinstance(col, tuple) else col for col in data.columns]

    # Print column names after flattening for debugging
    print("Column names after flattening:")
    display(data.columns)

    # Ensure 'Target_' column exists, recreating it at the top level if necessary using the flattened name
    if 'Target_' not in data.columns or data['Target_'].isnull().all(): # Check if 'Target_' is missing or all NaN
        # Check if 'Close_AAPL' exists after flattening
        if 'Close_AAPL' in data.columns:
            data['Target_'] = data['Close_AAPL'].shift(-1)
        else:
            print("Error: 'Close_AAPL' column not found in the DataFrame after flattening.")
            # Exit the execution of this cell if 'Close_AAPL' is not found
            raise SystemExit("Required column not found.")

        # Re-drop NaN values that might be created by the new Target_ column if it was missing
        data.dropna(inplace=True)


    # Create binary target based on whether tomorrow's close is higher than today's close
    # Access 'Close_AAPL' and 'Target_' at top level after flattening
    close_price_flattened = data['Close_AAPL']
    data['Price_Direction'] = (data['Target_'] > close_price_flattened).astype(int)

    # Drop the original 'Target_' column as we now have the binary 'Price_Direction'
    if 'Target_' in data.columns:
         data.drop('Target_', axis=1, inplace=True)


    # Inspect the data after creating the binary target
    print("Data after creating binary target variable:")
    display(data.head())
    display(data.tail())
    print("\nShape of data after creating binary target variable:", data.shape)
    print("Value counts for the binary target variable:")
    display(data['Price_Direction'].value_counts())


    # Define features (X) and target (y)
    # Exclude the 'Price_Direction' column itself from features
    # Features will be all columns except 'Price_Direction' and the flattened 'Target_'
    features = [col for col in data.columns if col not in ['Price_Direction', 'Target_']]
    X = data[features]
    y = data['Price_Direction']


    # Split data into training and testing sets (70/30 ratio)
    # Ensure we have enough samples after dropping NaNs
    if len(X) > 1: # Check if there's more than one sample to split
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y) # Use stratify for classification

        # Initialize and train the Random Forest Classifier model
        model = RandomForestClassifier(n_estimators=100, random_state=42) # Using 100 estimators as a starting point
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Evaluate the model
        accuracy = accuracy_score(y_test, y_pred)
        conf_matrix = confusion_matrix(y_test, y_pred)
        class_report = classification_report(y_test, y_pred)

        print(f"\nAccuracy: {accuracy:.2f}")
        print("\nConfusion Matrix:")
        display(conf_matrix)
        print("\nClassification Report:")
        print(class_report)

        # Report Feature Importance
        print("\nFeature Importance:")
        # Ensure feature_importances index matches the actual column names in X_train
        feature_importances = pd.Series(model.feature_importances_, index=X_train.columns).sort_values(ascending=False)
        display(feature_importances)

    else:
        print("\nNot enough data after creating binary target to perform train-test split.")


else:
    print("The 'data' DataFrame is empty. Please ensure the previous steps to load and process data were successful.")

Column names after flattening:


Index(['Close_AAPL', 'High_AAPL', 'Low_AAPL', 'Open_AAPL', 'Volume_AAPL',
       'MA_5_', 'MA_10_', 'MA_50_', 'Vol_Change_', 'Price_Change_',
       'Daily_Return_', 'Volatility_', 'Close_Lag1_', 'Close_Lag5_',
       'RSI_14_manual_', 'Target_'],
      dtype='object')

Data after creating binary target variable:


Close_AAPL  High_AAPL   Low_AAPL  Open_AAPL  Volume_AAPL  \
Date                                                                   
2020-03-13   67.369156  67.841763  61.305276  64.199074    370732000   
2020-03-16   58.702312  62.790941  58.166693  58.639296    322423600   
2020-03-17   61.283459  62.434671  57.778914  59.986825    324056000   
2020-03-18   59.783234  60.590297  57.468684  58.110943    300233600   
2020-03-19   59.325180  61.278611  58.799257  59.957743    271857200   

                MA_5_     MA_10_     MA_50_  Vol_Change_  Price_Change_  \
Date                                                                      
2020-03-13  65.589732  68.490313  74.025908    -0.114086       0.119808   
2020-03-16  64.428338  67.118549  73.747538    -0.130305      -0.128647   
2020-03-17  62.853963  66.234898  73.534911     0.005063       0.043970   
2020-03-18  61.459897  64.875979  73.280818    -0.073513      -0.024480   
2020-03-19  61.292668  63.709251  73.024383    -0.094514      -0.007662   

            Daily_Return_  Volatility_  Close_Lag1_  Close_Lag5_  \
Date                                                               
2020-03-13       0.119808     0.065253    60.161324    70.049667   
2020-03-16      -0.128647     0.073137    67.369156    64.509285   
2020-03-17       0.043970     0.074273    58.702312    69.155334   
2020-03-18      -0.024480     0.072621    61.283459    66.753563   
2020-03-19      -0.007662     0.072614    59.783234    60.161324   

            RSI_14_manual_  Price_Direction  
Date                                         
2020-03-13       44.954384                0  
2020-03-16       36.656287                1  
2020-03-17       40.196808                0  
2020-03-18       38.838027                0  
2020-03-19       38.411095                0

Close_AAPL   High_AAPL    Low_AAPL   Open_AAPL  Volume_AAPL  \
Date                                                                      
2022-12-22  130.539062  132.839269  128.633750  132.631963     77852100   
2022-12-23  130.173782  130.726619  127.982170  129.245801     63814900   
2022-12-27  128.367203  129.729560  127.073957  129.699945     69007800   
2022-12-28  124.428215  129.354401  124.260391  128.011792     85438400   
2022-12-29  127.952568  128.811438  126.096612  126.353282     75703700   

                 MA_5_      MA_10_      MA_50_  Vol_Change_  Price_Change_  \
Date                                                                         
2022-12-22  131.666443  136.105939  142.431838    -0.093984      -0.023773   
2022-12-23  131.143225  135.089108  142.216760    -0.180306      -0.002798   
2022-12-27  130.681216  133.661601  142.056421     0.081374      -0.013878   
2022-12-28  129.445229  131.743449  141.737864     0.238098      -0.030685   
2022-12-29  128.292166  130.400842  141.463382    -0.113938       0.028324   

            Daily_Return_  Volatility_  Close_Lag1_  Close_Lag5_  \
Date                                                               
2022-12-22      -0.023773     0.018734   133.717880   134.754440   
2022-12-23      -0.002798     0.018781   130.539062   132.789871   
2022-12-27      -0.013878     0.018176   130.173782   130.677246   
2022-12-28      -0.030685     0.019223   128.367203   130.608154   
2022-12-29       0.028324     0.020925   124.428215   133.717880   

            RSI_14_manual_  Price_Direction  
Date                                         
2022-12-22       36.101079                0  
2022-12-23       35.694643                0  
2022-12-27       33.675348                0  
2022-12-28       29.726642                1  
2022-12-29       36.860483                1


Shape of data after creating binary target variable: (706, 16)
Value counts for the binary target variable:


Price_Direction
1    365
0    341
Name: count, dtype: int64


Accuracy: 0.48

Confusion Matrix:


array([[41, 61],
       [50, 60]])


Classification Report:
              precision    recall  f1-score   support

           0       0.45      0.40      0.42       102
           1       0.50      0.55      0.52       110

    accuracy                           0.48       212
   macro avg       0.47      0.47      0.47       212
weighted avg       0.47      0.48      0.47       212


Feature Importance:


Volatility_       0.085723
RSI_14_manual_    0.083210
Daily_Return_     0.079163
Vol_Change_       0.073113
Price_Change_     0.071580
Volume_AAPL       0.070398
MA_50_            0.066071
Open_AAPL         0.061651
High_AAPL         0.061088
Close_Lag5_       0.060835
Close_AAPL        0.058897
Close_Lag1_       0.057763
Low_AAPL          0.057097
MA_5_             0.056983
MA_10_            0.056426
dtype: float64

## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

Here are some additional features that might help improve a stock price prediction model:

1.  **More Moving Averages:** Experiment with different window sizes (e.g., 20-day, 100-day, 200-day moving averages).
2.  **Other Technical Indicators:**
    *   Moving Average Convergence Divergence (MACD)
    *   Bollinger Bands
    *   Stochastic Oscillator
    *   Average True Range (ATR)
    *   Ichimoku Cloud
3.  **Lagged Prices:** Include the closing price from previous days as features.
4.  **Lagged Volume:** Include the trading volume from previous days.
5.  **Volatility Measures:**
    *   Parkinson Volatility
    *   Garman-Klass Volatility
6.  **Time-Based Features:**
    *   Day of the week
    *   Month of the year
    *   Quarter of the year
7.  **External Data:**
    *   News sentiment scores related to the stock or the overall market.
    *   Economic indicators (e.g., interest rates, inflation, GDP growth).
    *   Industry or sector specific indices.
    *   Related asset prices (e.g., commodities, currencies).
8.  **Fundamental Data:**
    *   Earnings per share (EPS)
    *   Price-to-earnings (P/E) ratio
    *   Dividend yield
    *   Revenue growth

# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



# Task
Explain why the accuracy of the Random Forest model is less than 50% and suggest ways to improve the model's performance.

## Feature engineering

### Subtask:
Explore creating more relevant features from the existing data or incorporating external data that might influence stock prices.


**Reasoning**:
To create more relevant features, I will implement the calculation of two additional technical indicators from the suggestions provided: MACD and Bollinger Bands. I will also add lagged volume features. After calculating these features, I will display the dataframe and check for NaN values.



In [29]:
# Calculate MACD
data['EMA_12'] = data['Close_AAPL'].ewm(span=12, adjust=False).mean()
data['EMA_26'] = data['Close_AAPL'].ewm(span=26, adjust=False).mean()
data['MACD'] = data['EMA_12'] - data['EMA_26']
data['Signal_Line'] = data['MACD'].ewm(span=9, adjust=False).mean()

# Calculate Bollinger Bands
data['Bollinger_Mid'] = data['Close_AAPL'].rolling(window=20).mean()
data['Bollinger_Std'] = data['Close_AAPL'].rolling(window=20).std()
data['Bollinger_Upper'] = data['Bollinger_Mid'] + (data['Bollinger_Std'] * 2)
data['Bollinger_Lower'] = data['Bollinger_Mid'] - (data['Bollinger_Std'] * 2)

# Lagged Volume
data['Volume_Lag1'] = data['Volume_AAPL'].shift(1)
data['Volume_Lag5'] = data['Volume_AAPL'].shift(5)

# Display the data after adding new features
print("Data after adding new features:")
display(data.head())
display(data.tail())

# Display NaN values per column
print("\nNumber of NaN values per column after adding new features:")
display(data.isnull().sum())

# Display shape of the DataFrame
print("\nShape of data after adding new features:", data.shape)

# Outline how external data could be integrated
print("\nPotential External Data Integration:")
print("- News Sentiment Scores: Could be added as a daily sentiment score based on news articles related to AAPL, capturing market mood.")
print("- Economic Indicators: Data like interest rates, inflation, or GDP growth could be included as macroeconomic features, potentially influencing overall market trends and thus stock prices.")
print("- Industry/Sector Indices: Including the performance of relevant technology sector indices could provide context on broader market movements affecting AAPL.")


Data after adding new features:


Close_AAPL  High_AAPL   Low_AAPL  Open_AAPL  Volume_AAPL  \
Date                                                                   
2020-03-13   67.369156  67.841763  61.305276  64.199074    370732000   
2020-03-16   58.702312  62.790941  58.166693  58.639296    322423600   
2020-03-17   61.283459  62.434671  57.778914  59.986825    324056000   
2020-03-18   59.783234  60.590297  57.468684  58.110943    300233600   
2020-03-19   59.325180  61.278611  58.799257  59.957743    271857200   

                MA_5_     MA_10_     MA_50_  Vol_Change_  Price_Change_  ...  \
Date                                                                     ...   
2020-03-13  65.589732  68.490313  74.025908    -0.114086       0.119808  ...   
2020-03-16  64.428338  67.118549  73.747538    -0.130305      -0.128647  ...   
2020-03-17  62.853963  66.234898  73.534911     0.005063       0.043970  ...   
2020-03-18  61.459897  64.875979  73.280818    -0.073513      -0.024480  ...   
2020-03-19  61.292668  63.709251  73.024383    -0.094514      -0.007662  ...   

               EMA_12     EMA_26      MACD  Signal_Line  Bollinger_Mid  \
Date                                                                     
2020-03-13  67.369156  67.369156  0.000000     0.000000            NaN   
2020-03-16  66.035795  66.727167 -0.691372    -0.138274            NaN   
2020-03-17  65.304667  66.323930 -1.019263    -0.314472            NaN   
2020-03-18  64.455215  65.839434 -1.384218    -0.528421            NaN   
2020-03-19  63.665979  65.356896 -1.690917    -0.760921            NaN   

            Bollinger_Std  Bollinger_Upper  Bollinger_Lower  Volume_Lag1  \
Date                                                                       
2020-03-13            NaN              NaN              NaN          NaN   
2020-03-16            NaN              NaN              NaN  370732000.0   
2020-03-17            NaN              NaN              NaN  322423600.0   
2020-03-18            NaN              NaN              NaN  324056000.0   
2020-03-19            NaN              NaN              NaN  300233600.0   

            Volume_Lag5  
Date                     
2020-03-13          NaN  
2020-03-16          NaN  
2020-03-17          NaN  
2020-03-18          NaN  
2020-03-19          NaN  

[5 rows x 26 columns]

Close_AAPL   High_AAPL    Low_AAPL   Open_AAPL  Volume_AAPL  \
Date                                                                      
2022-12-21  133.717880  135.060489  131.052411  131.279465     85928000   
2022-12-22  130.539062  132.839269  128.633750  132.631963     77852100   
2022-12-23  130.173782  130.726619  127.982170  129.245801     63814900   
2022-12-27  128.367203  129.729560  127.073957  129.699945     69007800   
2022-12-28  124.428215  129.354401  124.260391  128.011792     85438400   

                 MA_5_      MA_10_      MA_50_  Vol_Change_  Price_Change_  \
Date                                                                         
2022-12-21  132.509518  137.134613  142.547952     0.109711       0.023810   
2022-12-22  131.666443  136.105939  142.431838    -0.093984      -0.023773   
2022-12-23  131.143225  135.089108  142.216760    -0.180306      -0.002798   
2022-12-27  130.681216  133.661601  142.056421     0.081374      -0.013878   
2022-12-28  129.445229  131.743449  141.737864     0.238098      -0.030685   

            ...      EMA_12      EMA_26      MACD  Signal_Line  Bollinger_Mid  \
Date        ...                                                                 
2022-12-21  ...  136.872599  140.002744 -3.130145    -1.952106     140.594302   
2022-12-22  ...  135.898208  139.301730 -3.403522    -2.242389     139.664349   
2022-12-23  ...  135.017527  138.625586 -3.608059    -2.515523     138.862238   
2022-12-27  ...  133.994401  137.865706 -3.871305    -2.786680     138.161812   
2022-12-28  ...  132.522680  136.870336 -4.347656    -3.098875     137.414986   

            Bollinger_Std  Bollinger_Upper  Bollinger_Lower  Volume_Lag1  \
Date                                                                       
2022-12-21       5.507365       151.609033       129.579572   77432800.0   
2022-12-22       5.558801       150.781951       128.546747   85928000.0   
2022-12-23       5.718779       150.299796       127.424681   77852100.0   
2022-12-27       6.110273       150.382358       125.941265   63814900.0   
2022-12-28       6.826354       151.067694       123.762278   69007800.0   

            Volume_Lag5  
Date                     
2022-12-21   82291200.0  
2022-12-22   98931900.0  
2022-12-23  160156900.0  
2022-12-27   79592600.0  
2022-12-28   77432800.0  

[5 rows x 26 columns]


Number of NaN values per column after adding new features:


Close_AAPL          0
High_AAPL           0
Low_AAPL            0
Open_AAPL           0
Volume_AAPL         0
MA_5_               0
MA_10_              0
MA_50_              0
Vol_Change_         0
Price_Change_       0
Daily_Return_       0
Volatility_         0
Close_Lag1_         0
Close_Lag5_         0
RSI_14_manual_      0
Price_Direction     0
EMA_12              0
EMA_26              0
MACD                0
Signal_Line         0
Bollinger_Mid      19
Bollinger_Std      19
Bollinger_Upper    19
Bollinger_Lower    19
Volume_Lag1         1
Volume_Lag5         5
dtype: int64


Shape of data after adding new features: (705, 26)

Potential External Data Integration:
- News Sentiment Scores: Could be added as a daily sentiment score based on news articles related to AAPL, capturing market mood.
- Economic Indicators: Data like interest rates, inflation, or GDP growth could be included as macroeconomic features, potentially influencing overall market trends and thus stock prices.
- Industry/Sector Indices: Including the performance of relevant technology sector indices could provide context on broader market movements affecting AAPL.


## Model selection

### Subtask:
Experiment with other classification algorithms that might be better suited for this type of time-series data.


**Reasoning**:
Import the necessary classification models and define features and target variables, then split the data and train each model, finally evaluate each model's performance.



In [30]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pandas as pd

# Check if the data DataFrame from previous steps is not empty
if 'data' in locals() and not data.empty:

    # Define features (X) and target (y)
    # Exclude columns that would introduce data leakage ('Target_' and 'Price_Direction')
    # Also exclude the original price and volume columns if their lagged versions or indicators are used
    # Keep only the engineered features and relevant original features that do not cause leakage
    features = [col for col in data.columns if col not in ['Target_', 'Price_Direction']]
    X = data[features]
    y = data['Price_Direction']

    # Ensure we have enough samples after dropping NaNs
    if len(X) > 1:
        # Split data into training and testing sets (70/30 ratio)
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

        # Initialize and train Logistic Regression model
        print("Training Logistic Regression Model...")
        lr_model = LogisticRegression(random_state=42, solver='liblinear') # Using liblinear solver for convergence
        lr_model.fit(X_train, y_train)
        y_pred_lr = lr_model.predict(X_test)
        print("\n--- Logistic Regression Performance ---")
        print(f"Accuracy: {accuracy_score(y_test, y_pred_lr):.2f}")
        print("\nConfusion Matrix:")
        display(confusion_matrix(y_test, y_pred_lr))
        print("\nClassification Report:")
        print(classification_report(y_test, y_pred_lr))

        # Initialize and train K-Nearest Neighbors model
        print("\nTraining K-Nearest Neighbors Model...")
        knn_model = KNeighborsClassifier() # Using default k=5
        knn_model.fit(X_train, y_train)
        y_pred_knn = knn_model.predict(X_test)
        print("\n--- K-Nearest Neighbors Performance ---")
        print(f"Accuracy: {accuracy_score(y_test, y_pred_knn):.2f}")
        print("\nConfusion Matrix:")
        display(confusion_matrix(y_test, y_pred_knn))
        print("\nClassification Report:")
        print(classification_report(y_test, y_pred_knn))

        # Initialize and train Support Vector Machine model
        print("\nTraining Support Vector Machine Model...")
        # Using a linear kernel for simplicity and faster computation
        svm_model = SVC(kernel='linear', random_state=42)
        svm_model.fit(X_train, y_train)
        y_pred_svm = svm_model.predict(X_test)
        print("\n--- Support Vector Machine Performance ---")
        print(f"Accuracy: {accuracy_score(y_test, y_pred_svm):.2f}")
        print("\nConfusion Matrix:")
        display(confusion_matrix(y_test, y_pred_svm))
        print("\nClassification Report:")
        print(classification_report(y_test, y_pred_svm))

    else:
        print("\nNot enough data after creating binary target to perform train-test split.")

else:
    print("The 'data' DataFrame is empty. Please ensure the previous steps to load and process data were successful.")

Training Logistic Regression Model...


ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

**Reasoning**:
The previous code failed because there are NaN values in the features (X). Need to drop rows with NaN values before training the models.



In [ ]:
# Drop rows with NaN values
data.dropna(inplace=True)

# Define features (X) and target (y)
# Exclude columns that would introduce data leakage ('Target_' and 'Price_Direction')
# Also exclude the original price and volume columns if their lagged versions or indicators are used
# Keep only the engineered features and relevant original features that do not cause leakage
features = [col for col in data.columns if col not in ['Target_', 'Price_Direction']]
X = data[features]
y = data['Price_Direction']

# Ensure we have enough samples after dropping NaNs
if len(X) > 1:
    # Split data into training and testing sets (70/30 ratio)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

    # Initialize and train Logistic Regression model
    print("Training Logistic Regression Model...")
    lr_model = LogisticRegression(random_state=42, solver='liblinear') # Using liblinear solver for convergence
    lr_model.fit(X_train, y_train)
    y_pred_lr = lr_model.predict(X_test)
    print("\n--- Logistic Regression Performance ---")
    print(f"Accuracy: {accuracy_score(y_test, y_pred_lr):.2f}")
    print("\nConfusion Matrix:")
    display(confusion_matrix(y_test, y_pred_lr))
    print("\nClassification Report:")
    print(classification_report(y_test, y_pred_lr))

    # Initialize and train K-Nearest Neighbors model
    print("\nTraining K-Nearest Neighbors Model...")
    knn_model = KNeighborsClassifier() # Using default k=5
    knn_model.fit(X_train, y_train)
    y_pred_knn = knn_model.predict(X_test)
    print("\n--- K-Nearest Neighbors Performance ---")
    print(f"Accuracy: {accuracy_score(y_test, y_pred_knn):.2f}")
    print("\nConfusion Matrix:")
    display(confusion_matrix(y_test, y_pred_knn))
    print("\nClassification Report:")
    print(classification_report(y_test, y_pred_knn))

    # Initialize and train Support Vector Machine model
    print("\nTraining Support Vector Machine Model...")
    # Using a linear kernel for simplicity and faster computation
    svm_model = SVC(kernel='linear', random_state=42)
    svm_model.fit(X_train, y_train)
    y_pred_svm = svm_model.predict(X_test)
    print("\n--- Support Vector Machine Performance ---")
    print(f"Accuracy: {accuracy_score(y_test, y_pred_svm):.2f}")
    print("\nConfusion Matrix:")
    display(confusion_matrix(y_test, y_pred_svm))
    print("\nClassification Report:")
    print(classification_report(y_test, y_pred_svm))

else:
    print("\nNot enough data after creating binary target to perform train-test split.")

Training Logistic Regression Model...

--- Logistic Regression Performance ---
Accuracy: 0.53

Confusion Matrix:


array([[  4,  95],
       [  2, 105]])


Classification Report:
              precision    recall  f1-score   support

           0       0.67      0.04      0.08        99
           1       0.53      0.98      0.68       107

    accuracy                           0.53       206
   macro avg       0.60      0.51      0.38       206
weighted avg       0.59      0.53      0.39       206


Training K-Nearest Neighbors Model...

--- K-Nearest Neighbors Performance ---
Accuracy: 0.51

Confusion Matrix:


array([[43, 56],
       [45, 62]])


Classification Report:
              precision    recall  f1-score   support

           0       0.49      0.43      0.46        99
           1       0.53      0.58      0.55       107

    accuracy                           0.51       206
   macro avg       0.51      0.51      0.51       206
weighted avg       0.51      0.51      0.51       206


Training Support Vector Machine Model...
